simphony_osp/session/session.py

```diff
@@ -428,27 +436,41 @@ def from_label(
             The ontology entity.
         """
         results = set()
-        for identifier in self.iter_identifiers():
-            entity_labels = self.iter_labels(
-                entity=identifier,
-                lang=lang,
-                return_prop=False,
-                return_literal=False,
+
+        identifiers_and_labels = self.iter_labels(
+            lang=lang,
+            return_prop=False,
+            return_literal=False,
+            return_identifier=True,
+        )
```

Comment:
This is a super-good point. I assume you are referring to the fact that I could do something similar to
```python
def from_label(self, label: str, lang: Optional[str] = None, case_sensitive: bool = False) -> FrozenSet[OntologyEntity]:
    """[...]
    
    [...]
    """

    [...]

    identifiers = self._graph.triples((None, label_property, Literal(label)))  # let's ignore the languages for simplicity
    for identifier in identifiers:
        try:
            resuts.add(self.from_identifier(identifier))
        except KeyError:
            pass

    [...]

    return frozenset(results)
```

The default, in-memory store of RDFLib is implemented as a dictionary, so the lookup time for subjects should be `O(1)` using this approach. The approach on the PR loops over all labels in the graph, so it is `O(n)`, where `O(n)` is the number of labels in the graph. This means that **the approach that you suggest should be much faster**, at least in theory.

There is however an obvious drawback here: case-sensitivity. I do not expect the trick to work, for example, if `label == "amount"` but in the graph the triples are of the form `self._graph.triples(None, label_property, Literal("Amount"))`.  The solution to this problem is not as easy as try both "amount" and "Amount", because "amOunt" should also work. That is, if we want to use this approach, we have to compute all the combinations of possible capitalizations and do a lookup for each one. Things become harder with languages, because for each such combination also all allowed language strings have to be tried. This requires extra code and effort and if the user has installed only a few ontologies, the number of labels might be smaller than the number of capitalization and language combinations. But of course, if the number of installed ontologies grows, then at some point it starts to be worthwhile.

I think the tradeoff could be quite good for case-senstive cases, as the number of default label properties is only two (`skos:prefLabel` and `rdfs:label`). Just two checks, sounds awesome! But the language issue is still lurking in there: is it really worth also to do the check for all allowed languages (in terms of computational effort and human effort)?

The whole problem boils down to the fact that the RDFLib graph is a _"hash table"_ combining the hash of `(label, language)`, but even in the simple case of case-sensivity, what we can easily compute is just the hash of `label`.

Reply:
To add to this: maybe the sensible thing to do is to wait and see if some use case needs a large number of ontologies, and then invest the time in developing such an approach?

Reply:
Sorry for the misunderstanding then :|